### PR TITLE
C22/ Natalie Sen & Luqi Xie/ Viewing-Party

### PR DESCRIPTION
**Pull request by natsen1004**

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -159,12 +159,13 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
     
-    raise Exception("Test needs to be completed.")
+    #raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
+    assert MOVIE_TITLE_1 == updated_data["watched"][0]["title"]
```
Comment (apradoada) [inline tests/test_wave_01.py:166]: Great test! Just a couple of things here! Typically, we are checking to see if our new data is equal to a static variable. All this means is that our output should be on the left and the static variable should be on the right! Swap those for a more standard test.

Also, just to make sure our tests are super comprehensive, it might be better to assert that the entire movie is there as opposed to just the title. This may look something like this:

    assert updated_data["watched"][0] == {
            "title": MOVIE_TITLE_1,
            "genre": GENRE_1,
            "rating": RATING_1
        }

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -183,12 +184,13 @@ def test_moves_movie_from_watchlist_to_watched():
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
+    assert HORROR_1 in updated_data["watched"]
```
Comment (apradoada) [inline tests/test_wave_01.py:191]: This looks great! You could also check to se if the correct index is the movie you're looking for!

File: tests/test_wave_03.py [hunk 3]
```diff
@@ -55,12 +55,16 @@ def test_friends_unique_movies_not_duplicated():
...
     assert len(friends_unique_movies) == 3
 
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *************************************************************************************************
     # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
     # **************************************************************************************************
+    assert INTRIGUE_3 in friends_unique_movies
+    assert HORROR_1 in friends_unique_movies
+    assert FANTASY_4 in friends_unique_movies
```
Comment (apradoada) [inline tests/test_wave_03.py:64]: These look good! One other thing to think about is what this specific test is looking for. It's named `test_friends_unique_movies_not_duplicated`. This means we're checking to make sure there aren't any duplicates. The Arrange portion of this test ensures that there are two friends who have the same movie. Is there a way we could more robustly test for no duplicates?

File: tests/test_wave_05.py [hunk 4]
```diff
@@ -53,12 +53,15 @@ def test_new_genre_rec_from_empty_friends():
         ]
     }
 
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *********************************************************************
     # ****** Complete the Act and Assert Portions of these tests **********
     # *********************************************************************
+    recommendations = get_new_rec_by_genre(sonyas_data)
+    assert len(recommendations) == 0
```
Comment (apradoada) [inline tests/test_wave_05.py:61]: These are great! I love the way you implemented the Act portion of the test!

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,182 @@
...
-    pass
+
+    if title and genre and rating:
+        return {
+            "title": title,
+            "genre": genre,
+            "rating": rating
+        }
+    else: 
+        return None
```
Comment (apradoada) [inline viewing_party/party.py:12]: This is good, but I have one small tweak to suggest! When we look at functions like this, it is important to look at it in terms of a guard clause and the normal flow of the function. If this function runs as expected, a movie gets created. If either the title, genre or rating are missing, we simply return none. 

With this in mind, the guard clause is checking for a missing title, genre or rating, and the movie creation acts as the normal flow of the function. Typically, we would want the guard clause to be a stand alone if statement (no else) and we don't need to put everything else in the guard clause. This may look something like this:

     def create_movie(title, genre, rating):

         if not title or not genre or not rating:
             return None

        return {
            "title": title,
            "genre": genre,
            "rating": rating
        }
        
This is a small change, but this sort of format (guard clause, then regular flow of the function) is a more standard way of approaching it!

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,182 @@
...
+    user_data["watchlist"].append(movie)
+
+    return user_data
+
+def watch_movie(user_data, title):
+
+    for i in range(len(user_data["watchlist"])):
+        if title == user_data["watchlist"][i]["title"]:
+            user_data["watched"].append(user_data["watchlist"][i])
+            user_data["watchlist"].remove(user_data["watchlist"][i])
```
Comment (apradoada) [inline viewing_party/party.py:31]: This looks good! I wonder what it might look like if we used a for/each loop?

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,182 @@
...
+    return user_data
+        
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    watched_movies = user_data.get("watched", [])
```
Comment (apradoada) [inline viewing_party/party.py:42]: I love the way y'all used the `.get()` method here! It can sometimes be a little clunky pulling a specific list out of such a nested data structure, especially if you only use it once or twice, but the rest of your function really justifies its use!

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,182 @@
...
+
+    if not watched_movies:
+        return 0.0
+    
+    total_rating = 0.0
+    all_movies = len(watched_movies)
```
Comment (apradoada) [inline viewing_party/party.py:48]: This is a nitpick, but a variable called `all_movies` makes it seem like it's a list of all the movies as opposed to the number of movies. I would suggest something a little more specific like `num_movies`! Or even better, you could just access the length directly as opposed to through a variable!

     (`return total_rating/ len(watched_movies)`)

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,182 @@
...
+        return None
+    
+    all_genre = {}
+
+    for item in most_watched_genre:
+        genre = item.get("genre")
+        if genre in all_genre:
+            all_genre[genre] += 1
+        else:
+            all_genre[genre] = 1
```
Comment (apradoada) [inline viewing_party/party.py:71]: This is a great frequency dictionary creation! As you get more python, you can leverage the .get() method to cut out the if/else statements! You will likely make plenty of frequency dictionaries, so start getting comfortable with alternate ways of making one!

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,182 @@
...
+
+    most_watched = None
+    highest_count = 0
+
+    for genre, count in all_genre.items():
+        if count > highest_count:
+            highest_count = count
+            most_watched = genre
+
+    return most_watched
```
Comment (apradoada) [inline viewing_party/party.py:81]: The rest of this looks great!

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,182 @@
...
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+
+    user_watched = user_data["watched"]
+    user_watched_titles = set()
+
+    for movie in user_watched:
+        user_watched_titles.add(movie["title"])
```
Comment (apradoada) [inline viewing_party/party.py:94]: It looks like you never actually use `user_watched_titles` anywhere else in this function! You could achieve the same result without it!

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,182 @@
...
+    
+    friends_watched_titles = set()
```
Comment (apradoada) [inline viewing_party/party.py:96]: It's good practice to try and declare all your variables at the same time. I would move this line up to where you define `user_watched`

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,182 @@
...
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_titles.add(movie["title"])
+
+    unique_watched = []
```
Comment (apradoada) [inline viewing_party/party.py:102]: Same as above! Try grouping all your variable definitions/assignments together (excluding variables that are scoped to a loop or conditional).

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,182 @@
...
+
+    for movie in user_watched:
+        if movie["title"] not in friends_watched_titles:
+            unique_watched.append(movie)
+
+    return unique_watched
```
Comment (apradoada) [inline viewing_party/party.py:108]: One small tweak to make to this function overall would be that you confine everything to compare just the movie title as opposed to the whole movie. You also do this with the friends_watched_titles where you store just the titles and not the movies themselves. While this may be tempting, what if you have multiple movies with the same name but different genres, or ratings? In a situation like this, it is usually better to just compare the entire movie to other movies as opposed to just a singular aspect of that movie! I know that this is a little trickier because you used sets to avoid duplicates, but I wonder if there is a different way to avoid duplicates...?

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,182 @@
...
+    for movie in user_data["watched"]:
+        user_watched_titles.add(movie["title"])
+    
+    friends_watched_titles = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_titles.add(movie["title"])
+
+    friends_unique_watched = []
+    friends_unique_watched_titles = friends_watched_titles - user_watched_titles
```
Comment (apradoada) [inline viewing_party/party.py:122]: I love the way you used the set difference operator here!

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,182 @@
...
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie['title'] in friends_unique_watched_titles:
+                friends_unique_watched.append(movie)
+                friends_unique_watched_titles.remove(movie['title'])
```
Comment (apradoada) [inline viewing_party/party.py:128]: Your use of sets necessitates this second nested loop! I wonder if there is a way to achieve the same result without sets or a second set of loops?

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,182 @@
...
+    
+    return friends_unique_watched
+                
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+
+    recommended_movies = []
+    unique_movies = get_friends_unique_watched(user_data)
```
Comment (apradoada) [inline viewing_party/party.py:138]: Great job using a previous wave's functions!

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,182 @@
...
+    
+    for unique_movie in unique_movies:
+        for subscription in user_data["subscriptions"]:
+            if subscription == unique_movie["host"]:
```
Comment (apradoada) [inline viewing_party/party.py:142]: This works, but I think it's always a good idea to let python do the heavy lifting when it comes to the loops. If we just checked to see if the unique movie's host was in subscriptions, we could avoid having to explicitly create a nested loop!

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,182 @@
...
+
+    most_watched_genre = get_most_watched_genre(user_data)
+    friends_unique_watched_list = get_friends_unique_watched(user_data)
+    new_rec_by_genre = []
+
+    if not user_data["watched"]:
+        return new_rec_by_genre
+
+    for movie in friends_unique_watched_list:
+        if movie["genre"] in most_watched_genre:
```
Comment (apradoada) [inline viewing_party/party.py:160]: `most_watched_genre` should come in as a string, so it would be a better idea to just check and see if both values are equal to each other!

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,182 @@
...
+    
+    return recommended_movies
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+
+    most_watched_genre = get_most_watched_genre(user_data)
+    friends_unique_watched_list = get_friends_unique_watched(user_data)
```
Comment (apradoada) [inline viewing_party/party.py:153]: More great helper functions!

File: viewing_party/party.py [hunk 21]
```diff
@@ -1,23 +1,182 @@
...
+            new_rec_by_genre.append(movie)
+
+    return new_rec_by_genre
+
+def get_rec_from_favorites(user_data):
+    unique_watched = get_unique_watched(user_data)
+    fav_movies_list = user_data["favorites"]
+    recommended_movies = []
+
+    if not user_data["favorites"]:
```
Comment (apradoada) [inline viewing_party/party.py:170]: If you are going to create a variable that holds this information (fav_movies_list), then make sure you're using it consistently! If you decide not to use it, that works too as long as you are being consistent across the program. I think earlier you hold it in a list and then use it consistently in the function. Here, we just have a bit of a mismatch!

**Review by apradoada — COMMENTED** [review]
Overall, this looks great, y'all! It's getting a GREEN! 

I've left just a few things for y'all to take a look at, but nothing that I'm too worried about! All of my suggestions are mostly just so y'all can apply certain concepts to your code moving forward! You should be very proud of what you did here! You are welcome to make changes if you'd like, but it isn't required!
